homeassistant/components/mqtt/client.py

```diff
@@ -517,13 +622,18 @@ async def async_subscribe(
         # Only subscribe if currently connected.
         if self.connected:
             self._last_subscribe = time.time()
-            await self._async_perform_subscriptions(((topic, qos),))
+            await self._async_queue_subscriptions(((topic, qos),))
 
         @callback
         def async_remove() -> None:
             """Remove subscription."""
             self._async_untrack_subscription(subscription)
             self._matching_subscriptions.cache_clear()
+            max_qos: int | None = self._max_qos.get(topic)
+            new_max: int | None = (
+                self._max_qos_from_subscriptions(topic) if max_qos is not None else None
+            )
+            self._max_qos[topic] = new_max or 0
```

Comment:
Let's simplify this
```suggestion
            # Update self._max_qos
            subs = self._matching_subscriptions(topic)
            if subs:
                self._max_qos[topic] = max([sub.qos for sub in subs])
            else:
                del self._max_qos[topic]
```

Reply:
This part of the code this executed then the integration calls the unsub handler. We can set a new max qos, but there can be an ongoing race, hence we cannot remove the key here, but instead we remove it using `_pending_subscriptions_lock` in `_async_unsubscribe` which is thread safe.

Reply:
I moved all of the max qos code from the unsubscribe handler to _async_unsubscribe to make it thread safe